tikv-client/src/main/java/com/pingcap/tikv/predicates/TiKVScanAnalyzer.java

```diff
@@ -297,47 +296,106 @@ TiKVScanPlan buildIndexScan(
     }
   }
 
-  private Pair<Key, Key> buildTableScanKeyRangePerId(long id, IndexRange ir) {
-    Key startKey;
-    Key endKey;
+  private ScanRange buildTableScanKeyRangePerId(long id, IndexRange ir) {
+    boolean isDecimal=false;
+    ScanRange range=new ScanRange();
+    Key signedStartKey = null,signedEndKey=null,unsignedStartKey=null,unsignedEndKey=null;
     if (ir.hasAccessKey()) {
       checkArgument(
           !ir.hasRange(), "Table scan must have one and only one access condition / point");
-
       Key key = ir.getAccessKey();
       checkArgument(key instanceof TypedKey, "Table scan key range must be typed key");
-      TypedKey typedKey = (TypedKey) key;
-      startKey = RowKey.toRowKey(id, typedKey);
-      endKey = startKey.next();
+      Object obj = ((TypedKey) key).getValue();
+      if(obj instanceof Long){
+        signedStartKey=RowKey.toRowKey(id,new IntHandle((long)obj));
+        signedEndKey=signedStartKey.next();
+      }else if (obj instanceof BigDecimal){
```

Comment:
Have you checked whether obj is BigDecimal when range is unsignedRange?

Reply:
> Have you checked whether obj is BigDecimal when range is unsignedRange?

```java
 protected Object decodeNotNull(int flag, CodecDataInput cdi) {
    long ret;
    switch (flag) {
       ...
      case Codec.UINT_FLAG:
        ret = IntegerCodec.readULong(cdi);
        break;
        ...
    }
    if (isUnsignedLong()) {
      return unsignedValueOf(ret);
    }
    return ret;
  }
public boolean isUnsignedLong() {
    return tp == MySQLType.TypeLonglong && isUnsigned();
  }
```
Only unsignedLong will convert to bigDemical